### PR TITLE
Use SQL connection strings only and enable SPN user

### PR DESCRIPTION
**Pull request by LauJohansson**
This update makes it possible to connect to SQL database using Service Principal:

The two biggest changes are:
 - SqlServer class only use connection strings, and not self.properties - since properties actually can be contained in the connection string
 - Deployment of SQL server can now create a db user for the SPN using its AD. 

Requirements/ changes to the deployment pipeline:

The deploying SPN is SQL Server Admin (In order for it to login to the db in the server).
The SQL server has now an SystemAssignedIdentity so we can give it roles/permissions.
The SQL server identity has the role [Directory Reader.](https://docs.microsoft.com/en-us/azure/azure-sql/database/authentication-aad-service-principal?view=azuresql)

The attached link explains how to give the Directory Reader using Azure PowerShell. But since it is depricated in 2024, the Graph API is used instead.  The SQL server could perhaps have a lower-level permission for being able to create AD users in the SQL server. I have not investigated it further.

### CODE REVIEW

**Comment by LauJohansson** [comment]
The problem of given the SQL server "Directory reader" might fails since it requires P1 or P2 tenant license.

**Comment by LauJohansson** [comment]
IT IS possible to give the SQL server managed identity the Directory Reader role. However, I cannot find the way of assigning the role with az cli. Therefore, I try give the identity "Contributor" role.

**Comment by LauJohansson** [comment]
@christianhelle - can you help? I try to implement this step: https://docs.microsoft.com/en-us/azure/azure-sql/database/authentication-aad-service-principal-tutorial?view=azuresql#assign-directory-readers-permission-to-the-sql-logical-server-identity - what do I miss?

**Comment by christianhelle** [comment]
> @christianhelle - can you help? I try to implement this step: https://docs.microsoft.com/en-us/azure/azure-sql/database/authentication-aad-service-principal-tutorial?view=azuresql#assign-directory-readers-permission-to-the-sql-logical-server-identity - what do I miss?

I can try to help...

**Comment by LauJohansson** [comment]
> > @christianhelle - can you help? I try to implement this step: https://docs.microsoft.com/en-us/azure/azure-sql/database/authentication-aad-service-principal-tutorial?view=azuresql#assign-directory-readers-permission-to-the-sql-logical-server-identity - what do I miss?
> 
> I can try to help...

If you could do your magic using azure cli only it would be nice. The Azure Powershell is retiring in the future.

**Comment by LauJohansson** [comment]
> > @christianhelle - can you help? I try to implement this step: https://docs.microsoft.com/en-us/azure/azure-sql/database/authentication-aad-service-principal-tutorial?view=azuresql#assign-directory-readers-permission-to-the-sql-logical-server-identity - what do I miss?
> 
> I can try to help...

I have fixed the issue.

File: src/atc/sql/SqlServer.py [hunk 1]
```diff
@@ -37,28 +40,46 @@ def __init__(
...
-            f"database={database};queryTimeout=0;loginTimeout={self.timeout}"
+            f"database={database};"
+            f"queryTimeout=0;"
+            f"loginTimeout={self.timeout};"
+            f"driver={jdbc_driver};"
         )
 
-        self.username = username
-        self.password = password
+        if spnpassword and spnid and not password and not username:
```
[the remark this answers is not in the document]
Comment (LauJohansson) [reply src/atc/sql/SqlServer.py:53]: I have created a unittest

File: tests/cluster/sql/test_deliverysqlspn.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+from atc.config_master import TableConfigurator
+
+from . import extras
+from .DeliverySqlServerSpn import DeliverySqlServerSpn
+from .test_deliverysql import DeliverySqlServerTests
+
+
+class DeliverySqlServerSpnTests(DeliverySqlServerTests):
```
Comment (JeppeBlixen) [inline tests/cluster/sql/test_deliverysqlspn.py:8]: Should there be an integration here? I Only see a test setUpClass, and no test?

Comment (LauJohansson) [reply tests/cluster/sql/test_deliverysqlspn.py:8]: @JeppeBlixen  - it inhreits "DeliverySqlServerTests" and overwrites the setupclass to use the sqlserver with SPN credentials. See line 8.

**Review by JeppeBlixen — APPROVED** [review]

**Pull request merged**